loader-proto.c

```diff
@@ -241,21 +241,22 @@ shim_load_image(BOOLEAN BootPolicy, EFI_HANDLE ParentImageHandle,
 		}
 	}
 
-	in_protocol = 1;
-	efi_status = handle_image(SourceBuffer, SourceSize, &image->li,
-	                          &image->entry_point, &image->alloc_address,
-	                          &image->alloc_pages, parent_verified);
-	in_protocol = 0;
-	if (EFI_ERROR(efi_status))
-		goto free_image;
-
 	*ImageHandle = NULL;
 	efi_status = BS->InstallMultipleProtocolInterfaces(ImageHandle,
 					&SHIM_LOADED_IMAGE_GUID, image,
 					&EFI_LOADED_IMAGE_GUID, &image->li,
 					&gEfiLoadedImageDevicePathProtocolGuid,
 					image->loaded_image_device_path,
 					NULL);
+	if (EFI_ERROR(efi_status))
+		goto free_image;
+
+	in_protocol = 1;
+	efi_status = handle_image(SourceBuffer, SourceSize, &image->li,
+				  *ImageHandle, &image->entry_point,
+				  &image->alloc_address, &image->alloc_pages,
+				  parent_verified);
+	in_protocol = 0;
 	if (EFI_ERROR(efi_status))
 		goto free_alloc;
```

Comment:
By swapping the handle_image and the protocol installation, if handle_image fails (because of a security violation or other issues) the caller is still left with the image handle installed. The `free_alloc` section section below should uninstall and null the variable to avoid this:

```
free_alloc:
	BS->UninstallMultipleProtocolInterfaces(ImageHandle,
	                                &SHIM_LOADED_IMAGE_GUID, image,
	                                &EFI_LOADED_IMAGE_GUID, &image->li,
	                                &gEfiLoadedImageDevicePathProtocolGuid,
					image->loaded_image_device_path,
					NULL);
	*ImageHandle = NULL;
```

Reply:
Thanks for noticing this.  I'll have an updated patch soon.